Logic changes to join/quitStakePools

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->



- [x] Legacy joinStakePools automatically sets Abstain in Conway
- [x] check withdrawals and quitting (quitStakePools) with joinStakePools 

### Comments

In order to run
```
just conway-integration-tests-cabal-match STAKE_POOLS_
```

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

ADP-3261

paweljakubas committed:
Update `SHELLEY_STAKE_POOLS` tests for Conway

HeinrichApfelmus reviewed (commented):
Thank you! The integration tests look good to me, especially concerning the use of `getTransaction`.

That said, in the future, I would like to invest some effort into making the tests easier to understand and maintain — at the moment, separating test intent from programmatic details is hard. This would involve: Not using the `Link` module, exceptions instead of checking for `202` explicitly, encapsulating `getTransaction`. But for the purpose of testing functionality, this is good enough for now.

I would like to refactor the changes to the program logic, especially relating to the `inConway` boolean. For that purpose, @paweljakubas , could you split the pull request into two pull requests: 

1. A pull request that only contains the changes to the program logic.
2. A pull request that uses pull request 1 as base and only contains the tests.

Thanks!

paweljakubas commented on lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -767,7 +767,7 @@ mkUnsignedTransaction networkId stakeCred ctx selection = do
                             , "action"
                             ]
-            let certs = L.nub $ votingCerts <> delegCerts
+            let certs = L.nub $ delegCerts <> votingCerts
```
This one is needed as Registration cert needs to be FIRST...Otherwise node can protest

HeinrichApfelmus replied:
I don't quite understand what's going on, because I would expect both `votingCerts` and `delegCerts` to contain the initial registration certificate. 🤔  That's actually my main reason for starting the refactoring.

paweljakubas replied:
hmmm....I found one case where this was crucial and got node error....but it was without your refactoring (ie., before I rebased aginst your PR). Anyway, it does not change the picture if we have `votingCerts <> delegCerts` or `delegCerts <> votingCerts`.

HeinrichApfelmus replied:
> I found one case where this was crucial and got node error.

Yes, I believe you — I just don't understand why, and that makes me suspicious that there could be another logic bug in the code. 🤔

HeinrichApfelmus replied:
Found it:

`joinStakePoolDelegationAction` must return `Tx.VoteRegisteringKey` instead of `Tx.Vote` in the voting action in order to for the second definition of certs to work when the key is not registered yet. I have fixed it.

paweljakubas committed:
voting/deleg/withdrawal integration tests

HeinrichApfelmus commented on lib/wallet/src/Cardano/Wallet/IO/Delegation.hs:
```diff
@@ -122,13 +122,18 @@ selectCoinsForJoin ctx pools poolId poolStatus = do
+        Left _ -> pure False
+        Right _ -> pure True
+
+    (action,_) <- WD.joinStakePoolDelegationAction @s
```
🤔 Shouldn't the voting certificate also be part of the transaction that we select coins for?

paweljakubas replied:
@Anviking @jonathanknowles   ????

Anviking replied:
So this endpoint is used by Daedalus for hardware wallet support.

If we want voting supported there, we should ensure the certificate is accounted for in the coin-selection, yes, but also that it is returned in the response. And then Daedalus would need to be updated to add that certificate to the tx.

HeinrichApfelmus replied:
> And then Daedalus would need to be updated to add that certificate to the tx.

Dang. We can select the coins for it, but we don't assemble the transaction — it looks like Daedalus can't even vote to abstain in the presence of a hardware wallet.

HeinrichApfelmus commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2677,6 +2678,12 @@ alreadyVoted walletState =
+    :: DBLayer IO s
+    -> IO Bool
+haveWeVoted DBLayer{..} = do
+    atomically (alreadyVoted walletState)
```
When rebasing on #4477, please use `alreadyVoted` directly in the pure functions `joinStakePoolDelegationAction` and `Cardano.Wallet.Delegation`; this function will become obsolete.

HeinrichApfelmus commented on lib/wallet/src/Cardano/Wallet/Delegation.hs:
```diff
@@ -172,10 +175,11 @@ joinStakePoolDelegationAction
 joinStakePoolDelegationAction
     tr DBLayer{..} currentEpochSlotting
-        knownPools poolId poolStatus = do
+        knownPools poolId poolStatus inConway = do
```
```suggestion
        knownPools poolId poolStatus era = do
```

When rebasing on #4477, please pass a value of type `era :: Write.RecentEra era` to the pure function instead.

HeinrichApfelmus commented on lib/wallet/src/Cardano/Wallet/Delegation.hs:
```diff
@@ -172,10 +175,11 @@ joinStakePoolDelegationAction
     -> PoolLifeCycleStatus
-    -> IO Tx.DelegationAction
+    -> Bool
+    -> IO (Tx.DelegationAction, Maybe Tx.VotingAction)
```
```suggestion
    -> IO (Tx.DelegationAction, Maybe Tx.VotingAction)
```

Return a tuple is not pretty, but it's ok.

HeinrichApfelmus commented on lib/wallet/src/Cardano/Wallet/Delegation.hs:
```diff
@@ -289,11 +299,16 @@ quitStakePool netLayer db timeInterpreter = do
+    :: WalletDelegation
+    -> Withdrawal
+    -> Coin
+    -> Bool
```
```suggestion
```

Isn't this piece of information contained in the `WalletDelegation` argument? 🤔 

```hs
voting = case last_ of
   Voting{} → True
   DelegatingVoting{} → True
   _ → False
```

HeinrichApfelmus replied:
Probably, but let's leave it like this for now — the argument is localized to the `Cardano.Wallet.Delegation` module now.

HeinrichApfelmus reviewed (commented):
@paweljakubas Thanks for splitting!

Could you rebase the logic on #4477 ? I have added additional comments — I would like to obtain a strict separation between `IO` (reading data from mutable sources) and pure functions (performing computation on that data).

HeinrichApfelmus committed:
Import qualified Cardano.Wallet.IO.Delegation shorthand

HeinrichApfelmus committed:
Add vote to `Abstain` to `joinStakePoolDelegationAction`

HeinrichApfelmus committed:
Change `guardQuit` to allow quitting even when voting

HeinrichApfelmus reviewed (approved):
I have applied my review suggestions — I'm happy now. 😊 Good to merge.

paweljakubas committed:
Voting/delegation/withdrawal integration tests (#4476)

<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
✓ Self-review your changes to make sure nothing unexpected slipped
through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->

- [x] Voting/revoting after delegation
- [x] Delegation/re-delegation after voting
- [x] Delegation/quitting after voting
- [x] Delegation together with voting
- [x] Quitting after voting
- [x] Withdrawals after voting
- [x] Withdrawals before voting

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number
adp-3261

<!-- Reference the Jira/GitHub issue that this PR relates to, and which
requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->
